### PR TITLE
feat!: refactor setup

### PR DESCRIPTION
**Pull request by AlphaKeks**
## 📃 Summary

As discussed in #76, the exposed options have been refactored to allow more customization of the side buffers. We now provide:
- `common`: defines options common to both buffers
- `left`/`right`: defines options for the side buffer directly

### CODE REVIEW

**Comment by AlphaKeks** [comment]
I will open another PR with the text-color feature once this is merged, if that's okay with you.

**Comment by shortcuts** [comment]
Thanks a lot for the PR! I'll give it a full pass once I'm done with https://github.com/shortcuts/no-neck-pain.nvim/pull/79 as I'd love to fix a bug before

File: README.md [hunk 1]
```diff
@@ -72,52 +72,62 @@ require("no-neck-pain").setup({
...
-        -- When `true`, the side buffers will be named `no-neck-pain-left` and `no-neck-pain-right` respectively.
-        showName = false,
-        -- The buffer options when creating the buffer.
-        options = {
-            -- Buffer-scoped options, below are the default values, but any `vim.bo` options are valid and will be forwarded to the buffer creation.
+        left = {
+            enabled = true,
+            -- Hex color for setting the background color of the NNP buffer as well as some other
+            -- highlight groups to make it look clean
+            color = nil,
```
Comment (shortcuts) [inline README.md:82]:
```suggestion
            backgroundColor = nil,
```

so it's ready for when we add the `textColor` one and there's no more breaking changes

File: lua/no-neck-pain/config.lua [hunk 2]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
     -- When `true`, disabling NNP kills every split/vsplit buffers except the main NNP buffer.
     killAllBuffersOnDisable = false,
     -- Options related to the side buffers.
+    -- When `true`, the side buffers will be named `no-neck-pain-left` and `no-neck-pain-right` respectively.
+    showBufferNames = false,
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:21]: actually as we rename it, it should be 
```suggestion
    setBufferNames = false,
```

File: lua/no-neck-pain/config.lua [hunk 3]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
...
-        left = true,
-        -- When `false`, the `right` padding buffer won't be created.
-        right = true,
-        -- When `true`, the side buffers will be named `no-neck-pain-left` and `no-neck-pain-right` respectively.
-        showName = false,
-        -- The buffer options when creating the buffer.
-        options = {
-            -- Buffer-scoped options, below are the default values, but any `vim.bo` options are valid and will be forwarded to the buffer creation.
+        left = {
+            enabled = true,
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:24]:
```suggestion
            -- When `false`, the buffer won't be created.
            enabled = true,
```

Same for the `right` one

File: lua/no-neck-pain/config.lua [hunk 4]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
...
+            -- Hex color for setting the background color of the NNP buffer as well as some other
+            -- highlight groups to make it look clean
+            color = nil,
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:27]:
```suggestion
            backgroundColor = nil,
```

so it's already named for the `textColor` one is introduced

File: lua/no-neck-pain/config.lua [hunk 5]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
...
-        right = true,
-        -- When `true`, the side buffers will be named `no-neck-pain-left` and `no-neck-pain-right` respectively.
-        showName = false,
-        -- The buffer options when creating the buffer.
-        options = {
-            -- Buffer-scoped options, below are the default values, but any `vim.bo` options are valid and will be forwarded to the buffer creation.
+        left = {
+            enabled = true,
+            -- Hex color for setting the background color of the NNP buffer as well as some other
+            -- highlight groups to make it look clean
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:26]: can you keep the previous description please?

File: lua/no-neck-pain/config.lua [hunk 6]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
...
+            color = nil,
             bo = {
+                -- buffer-scoped options
+                -- Note: any `vim.bo` options will work here
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:30]:
```suggestion
            -- buffer-scoped options
            -- Note: any `vim.bo` options will work here
            bo = {
```

File: lua/no-neck-pain/config.lua [hunk 7]
```diff
@@ -17,52 +17,62 @@ NoNeckPain.options = {
...
                 bufhidden = "hide",
                 modifiable = false,
                 buflisted = false,
-                swapfile = false,
+                swapfile = false
             },
-            -- Window-scoped options, below are the default values, but any `vim.wo` options are valid and will be forwarded to the buffer creation.
             wo = {
+                -- window-scoped options
+                -- Note: any `vim.wo` options will work here
```
Comment (shortcuts) [inline lua/no-neck-pain/config.lua:40]:
```suggestion
            -- window-scoped options
            -- Note: any `vim.wo` options will work here
            wo = {
```

File: lua/no-neck-pain/main.lua [hunk 8]
```diff
@@ -59,23 +59,27 @@ local function createBuf(name, cmd, padding, moveTo)
...
 
-    if _G.NoNeckPain.config.buffers.background.colorCode ~= nil then
-        D.print("CreateWin: setting `colorCode` for buffer " .. id)
+    D.print(string.format("CreateWin: setting color `%s` for left buffer (`%s`)", _G.NoNeckPain.config.buffers.left.color, id))
+    C.init(id, _G.NoNeckPain.config.buffers.left.color)
 
-        C.init(id, _G.NoNeckPain.config.buffers.background.colorCode)
-    end
+    D.print(string.format("CreateWin: setting color `%s` for right buffer (`%s`)", _G.NoNeckPain.config.buffers.right.color, id))
+    C.init(id, _G.NoNeckPain.config.buffers.right.color)
```
Comment (shortcuts) [inline lua/no-neck-pain/main.lua:82]: you can move the `D.print` statement inside the init method so it's less redundant here

File: lua/no-neck-pain/util/color.lua [hunk 9]
```diff
@@ -30,31 +30,47 @@ function C.matchIntegrationToHexCode(colorCode)
...
+
+    local defaultBackground = vim.api.nvim_get_hl_by_name("Normal", true).background
+
+    -- check if the user has a transparent background or not
+    if defaultBackground == nil then
+        defaultBackground = "NONE"
+    else
+        -- if it's not transparent, get the user's current background color
+        defaultBackground = string.format("#%06X", defaultBackground)
+    end
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:46]: love it!

**Review by shortcuts — COMMENTED** [review]
thanks so much for taking the time!!

**Comment by AlphaKeks** [comment]
Nice, thank you! I think that's it, if I forgot anything please tell me.

File: lua/no-neck-pain/main.lua [hunk 10]
```diff
@@ -59,23 +59,24 @@ local function createBuf(name, cmd, padding, moveTo)
...
+    if _G.NoNeckPain.config.showBufferNames then
         vim.api.nvim_buf_set_name(0, "no-neck-pain-" .. name)
     end
 
-    for scope, _ in pairs(_G.NoNeckPain.config.buffers.options) do
-        for key, value in pairs(_G.NoNeckPain.config.buffers.options[scope]) do
-            vim[scope][key] = value
-        end
+    -- there is probably a cleaner way to do this but it's kind of annoying
+    -- since the tables contain multiple different kinds of values
```
Comment (shortcuts) [inline lua/no-neck-pain/main.lua:67]:
```suggestion
```

I like your solution, clear an concise to me so no need to justify it

File: lua/no-neck-pain/main.lua [hunk 11]
```diff
@@ -59,23 +59,24 @@ local function createBuf(name, cmd, padding, moveTo)
...
+    end
 
-    if _G.NoNeckPain.config.buffers.background.colorCode ~= nil then
-        D.print("CreateWin: setting `colorCode` for buffer " .. id)
+    vim.cmd(moveTo)
 
-        C.init(id, _G.NoNeckPain.config.buffers.background.colorCode)
-    end
+    C.init(id, _G.NoNeckPain.config.buffers.left.color)
+    C.init(id, _G.NoNeckPain.config.buffers.right.color)
```
Comment (shortcuts) [inline lua/no-neck-pain/main.lua:79]:
```suggestion
    C.init(id, _G.NoNeckPain.config.buffers[name].color)
```

The function is called for each side so we can just do it like this

File: lua/no-neck-pain/util/color.lua [hunk 12]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
     return colorCode
 end
 
--- creates an highlight group `NoNeckPain` with the given `colorCode` and assign it to the side buffer of the given `id`.
+-- creates an highlight group `NNPBuffers` with the given `color` and assign it to the side buffer of the given `id`.
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:33]:
```suggestion
-- creates an highlight group `NNPBuffers` with the given `backgroundColor` and assign it to the side buffer of the given `id`.
```

so it's ready for the `textColor` feature

File: lua/no-neck-pain/util/color.lua [hunk 13]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
...
 -- `cmd` is used instead of native commands for backward compatibility with Neovim 0.7
-function C.init(win, colorCode)
-    local groupName = "nnpcolorname"
+function C.init(win, color)
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:35]:
```suggestion
function C.init(win, backgroundColor)
```

File: lua/no-neck-pain/util/color.lua [hunk 14]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
...
+    D.print(string.format("CreateWin: setting color `%s` for right buffer (`%s`)", color, win))
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:36]:
```suggestion
    D.print(string.format("CreateWin: setting color `%s` for buffer `%s`", backgroundColor, win))
```

File: lua/no-neck-pain/util/color.lua [hunk 15]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
...
+
+    -- check if the user has a transparent background or not
+    if defaultBackground == nil then
+        defaultBackground = "NONE"
+    else
+        -- if it's not transparent, get the user's current background color
+        defaultBackground = string.format("#%06X", defaultBackground)
+    end
+
+    color = color or defaultBackground
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:50]:
```suggestion
    backgroundColor = backgroundColor or defaultBackground
```

File: lua/no-neck-pain/util/color.lua [hunk 16]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
...
+        string.format(
+        	[[
+Color.init: initializing colors:
+- groupName: `%s`
+- window: `%s`
+- color: `%s`
+            ]],
+            groupName,
+            win,
+            color
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:62]:
```suggestion
        	[[
Color.init: initializing colors:
- groupName: `%s`
- window: `%s`
- backgroundColor: `%s`
            ]],
            groupName,
            win,
            backgroundColor
```

File: lua/no-neck-pain/util/color.lua [hunk 17]
```diff
@@ -30,31 +30,49 @@ function C.matchIntegrationToHexCode(colorCode)
...
     vim.cmd(
         string.format(
             [[highlight! %s guifg=%s guibg=%s]],
             groupName,
-            colorCode,
-            colorCode,
-            colorCode,
-            colorCode
+            color,
+            color
```
Comment (shortcuts) [inline lua/no-neck-pain/util/color.lua:72]:
```suggestion
            backgroundColor,
            backgroundColor
```

**Review by shortcuts — COMMENTED** [review]
awesome! Last round of review, otherwise it seems really good to me.

Could you also run `make lint` and `make documentation`?

**Comment by AlphaKeks** [comment]
I ran the lint check, but the documentation check didn't work. Can't really tell why though...

```
nvim --headless --noplugin -u ./scripts/minimal_init.lua -c "lua require('mini.doc').generate()" -c "qa!"
Error detected while processing /home/alphakeks/Projects/no-neck-pain.nvim.git/refactor-options/scripts/minimal_init.lua:
E5113: Error while calling lua chunk: ...-pain.nvim.git/refactor-options/scripts/minimal_init.lua:11: module 'mini.test' not found:
        no field package.preload['mini.test']
        no file './mini/test.lua'
        no file '/usr/share/luajit-2.1.0-beta3/mini/test.lua'
        no file '/usr/local/share/lua/5.1/mini/test.lua'
        no file '/usr/local/share/lua/5.1/mini/test/init.lua'
        no file '/usr/share/lua/5.1/mini/test.lua'
        no file '/usr/share/lua/5.1/mini/test/init.lua'
        no file './mini/test.so'
        no file '/usr/local/lib/lua/5.1/mini/test.so'
        no file '/usr/lib/lua/5.1/mini/test.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './mini.so'
        no file '/usr/local/lib/lua/5.1/mini.so'
        no file '/usr/lib/lua/5.1/mini.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
        [C]: in function 'require'
        ...-pain.nvim.git/refactor-options/scripts/minimal_init.lua:11: in main chunkError detected while processing command line:
E5108: Error executing lua [string ":lua"]:1: module 'mini.doc' not found:
        no field package.preload['mini.doc']
        no file './mini/doc.lua'
        no file '/usr/share/luajit-2.1.0-beta3/mini/doc.lua'
        no file '/usr/local/share/lua/5.1/mini/doc.lua'
        no file '/usr/local/share/lua/5.1/mini/doc/init.lua'
        no file '/usr/share/lua/5.1/mini/doc.lua'
        no file '/usr/share/lua/5.1/mini/doc/init.lua'
        no file './mini/doc.so'
        no file '/usr/local/lib/lua/5.1/mini/doc.so'
        no file '/usr/lib/lua/5.1/mini/doc.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './mini.so'
        no file '/usr/local/lib/lua/5.1/mini.so'
        no file '/usr/lib/lua/5.1/mini.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
        [C]: in function 'require'
        [string ":lua"]:1: in main chunk%                                                                                                                               
```

**Comment by shortcuts** [comment]
Ah yes you might need to clone it, does it work if you run `make deps` before?

I should definitely improve that

**Comment by AlphaKeks** [comment]
Nope. Doesn't seem to exist.
``` 
make: *** No rule to make target 'deps'.  Stop.
```

**Comment by shortcuts** [comment]
> Nope. Doesn't seem to exist.
> 
> ``` 
> 
> make: *** No rule to make target 'deps'.  Stop.
> 
> ```
> 
> 
> 
> Edit: what should I clone to run it?

Ah it's because your branch is not up to date with `main`, either merge it or use the previous name: `make deps/mini.nvim`

**Comment by AlphaKeks** [comment]
It ran now, but it's saying there is a duplicate tag. When looking at the file I can only find it once though.

```
nvim --headless --noplugin -u ./scripts/minimal_init.lua -c "lua require('mini.doc').generate()" -c "qa!"
Error detected while processing command line:
E5108: Error executing lua deps/mini.nvim/lua/mini/doc.lua:402: Vim(helptags):E154: Duplicate tag "NoNeckPain.disable()" in file doc/refactor-options.txt
stack traceback:
        [C]: in function 'cmd'
        deps/mini.nvim/lua/mini/doc.lua:402: in function 'write_post'
        deps/mini.nvim/lua/mini/doc.lua:583: in function 'generate'
        [string ":lua"]:1: in main chunk%                                                                                                                               
```

**Comment by shortcuts** [comment]
hey! taking a look locally

**Comment by shortcuts** [comment]
I'm trying to think of a way to make the option also work globally, as I don't want user to have to repeat options if they want to apply it to both buffers

**Comment by shortcuts** [comment]
I think I came up with something satisfying, will push it here asap

**Comment by shortcuts** [comment]
hey @AlphaKeks, what do you think of this version?

**Review by AlphaKeks — COMMENTED** [review]
Love it! This will make implementing new features way easier in the future.

**Comment by shortcuts** [comment]
Not sure why it fails on 0.7.2, there's no changes on this side 🤔

**Comment by shortcuts** [comment]
I guess we are good to go here, let's iterate! Thanks A LOT for the contribution! :D

**Comment by AlphaKeks** [comment]
Great! No problem :)

**Pull request merged**